### ISSUE TITLE
Injecting `this` into pure component

### ISSUE THREAD

**Issue by riwu** [acdlite/recompose#587]
```
const App = () => (
  <View>
    <TextInput
      onChangeText={(text) => {
        this.text = text;
      }}
    />
    <Button onPress={() => console.log(this.text)} />
  </View>
);
```
In certain use cases, there's no need for a changing value to be a state, hence the use of `this` would prevent unnecessary re-rendering of the entire `App` component when the value is changed. 

Is it possible for `recompose` to support this? Or is the only way to turn the component into a `class`?

**Comment by pokorson** [comment]
@riwu you can't use `this` inside stateless functional components and it's how React work.
You can achieve this with class component but I believe it's anti-pattern in most cases

**Comment by istarkov** [comment]
> Or is the only way to turn the component into a class?

Yes, there are no other way to have some kind of `this` without instance created somewhere.

(It's possible to use a very dirty hack using component ref's and use a DOM element as data holder (as like as we did in the past with jQuery) but I never said this and don't recommend at all ;-))

**Comment by riwu** [comment]
@pokorson in the use case I mentioned above, would it be appropriate to use class component to prevent re-rendering? Or do you have suggestions on another pattern?

We could use `shouldUpdate` with a shallow comparison of all props except the ones used for the `TextInput`, but this comparison would still be 1 layer deeper compare to the default ---`currentProps === nextProps`, which is not a big deal if the component has only a few props though.

**Comment by istarkov** [comment]
> We could use shouldUpdate with a shallow comparison of all props except the ones used for the TextInput

Bad idea with high probability will cause a problem in a future. 
It's hard to find why component is not updated if update will be needed. After few similar problems with a lot of time spending in debug we decided not to use any kind of such optimisations except `pure` or naked shallowEqual on all props state.

In most cases just use state - it will be a hard to see any performance problems. Also for now it's not a problem to use `this`, even it will be an antipattern in the future some kind of `state` updater without rerender will be provided. As like as in reasonml now.

**Comment by riwu** [comment]
> Bad idea with high probability will cause a problem in a future.
> It's hard to find why component is not updated if update will be needed.
You could do something like `Object.entries(props).some(([key, value]) => key !== 'text' && value !== nextProps[key])` so that only the desired prop is ignored and other prop additions are checked by default.

I agree that this might be an unnecessary complication as opposed to just using `class` or let it re-render with `state` (which is indeed not causing any performance issues, at least on a modern device, so perhaps this is a premature/over optimisation)

**Comment by istarkov** [comment]
At the beginning of my React life every my component had `shouldComponentUpdate`, I used `this` whenever possible.
Now I almost don't use any optimisations at components level (sometimes at controls) and all is working fast and fine ;-) 
Having that most 3rd party controls are optimised my components code is not a big part of react tree, so just using few patterns - like never pass anonymous functions as callbacks into controls I have a good performance without any additional steps.
For 99% cases don't optimise at all is a good strategy, easier to support, less code.

**Comment by wuct** [comment]
I'm agree with @istarkov. Because we have 16 ms to update a view in the front-end, we can use this budget to trade simpler code. In addition, we can always come back later to optimize performance when we actually hit the 16-ms wall.

**Comment by Rusfighter** [comment]
@wuct @istarkov That might be not a problem on the browser, but when you are developing for mobile you really should squeeze those re-renders. 

BUT, since i am not fan of classes at all in JS but factories, I was wondering of something like this could work:

```
const viewFactor = () => {
  // static fields
  let staticText;

  return (props) => (
    <View>
      <TextInput
        onChangeText={(text) => {
          staticText = text;
        }}
      />
      <Button onPress={() => console.log(staticText)} />
    </View>
  )
}

export const someView = compose(
  ...
)(viewFactor());
```
Or will this create memory leaks/not properly rerenders?

**Comment by istarkov** [comment]
This is not reusable. You can use such component only if its a singleton or created inside a class :-)

**Comment by istarkov** [comment]
Imagine you need 2 someView on one page

**Comment by riwu** [comment]
Another common use case of `this` is to keep a `ref` to a component, which is unavoidable in some cases (eg. audio/video components).

> in the future some kind of state updater without rerender will be provided

I suppose `class` still has its use until this happens.

**Comment by Rusfighter** [comment]
Then we need to stick with classes for now i think
